lib/npg_pipeline/pluggable.pm

```diff
@@ -701,32 +702,55 @@ sub _run_spider {
   return;
 }
 
-#####
-# Attempts to create in the analysis directory a hard link to
-# or, if unsuccessful, a copy of the log file of the pipeline.
-# All errors are captured. Best effort.
-#
-sub _create_log_link {
-  my $self = shift;
+# Attempts to link the pipeline log file to the analysis directory
+sub _copy_log_to_analysis_dir {
+  my ($self) = @_;
+  $self->_tolerant_persist_file_to_analysis_dir(
+    $self->log_file_path, $self->log_file_name
+  );
+  return;
+}
 
-  (-e $self->analysis_path) or return;
+sub _save_product_conf_to_analysis_dir {
+  my ($self) = @_;
+  # First check for unusual file names:
+  my ($filename, $dirs, $suffix) = fileparse(
+    $self->product_conf_file_path,
+    qr/[.][^.]*/xms
+  );
+  if ($filename ne 'product_release') {
+    $filename = 'product_release_'.$filename;
+  }
+  $filename = $filename.'_'.$self->random_string.$suffix;
 
-  my $link = catfile($self->analysis_path, $self->log_file_name);
-  if (!-e $link) {
-    try {
-      my $linked = 1;
-      if (not link $self->log_file_path, $link) {
-        # If cannot hard link, just copy.
-        $linked = copy $self->log_file_path, $link;
-      }
-      $linked and $self->info("Created link or copy $link for the log file");
-    } catch {
-      $self->warn("Error creating a link or copy: $_");
-    };
+  $self->_tolerant_persist_file_to_analysis_dir(
+    $self->product_conf_file_path,
+    $filename
+  );
+  return $filename;
+}
+
+# Copies a file to analysis_path, and optionally
+# renames the file in its destination
+# Errors are captured where possible to prevent job resubmission in the
+# automatic pipeline
+sub _tolerant_persist_file_to_analysis_dir {
+  my ($self, $source_file, $override_name) = @_;
+
+  my $analysis_path = $self->analysis_path;
+  (-e $analysis_path) or return;
+  my $target_file;
+  if ($override_name) {
+    $target_file = catfile($analysis_path, $override_name);
   } else {
-    $self->warn("Will not overwrite existing file $link");
+    $target_file = catfile($analysis_path, $source_file);
```

Comment:
This path through the code is not used, neither it is tested. I believe it will create an incorrect target file path since the analysis path is concatenated with the path of the source file.

Reply:
Turns out it is used earlier in the tests where no log_file_name has been set for the pluggable instance. You're right about the catfile being suspect.